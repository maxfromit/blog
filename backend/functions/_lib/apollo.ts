import {
  ApolloClient,
  ApolloLink,
  HttpLink,
  concat,
  gql,
} from '@apollo/client/core'
import { InMemoryCache } from '@apollo/client/cache'

const httpLink = new HttpLink({
  fetch,
  uri: process.env.HASURA_GRAPHQL_GRAPHQL_URL,
})

const headersMiddleware = new ApolloLink((operation, forward) => {
  operation.setContext(({ headers = {} }) => ({
    headers: {
      ...headers,
      'x-hasura-admin-secret': process.env.HASURA_GRAPHQL_ADMIN_SECRET,
    },
  }))

  return forward(operation)
})

const link = concat(headersMiddleware, httpLink)
const cache = new InMemoryCache()

const client = new ApolloClient({ link, cache })

export { client as apollo, gql }
