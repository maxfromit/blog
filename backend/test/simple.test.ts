import { beforeAll, describe, expect, test } from 'vitest'
import { apollo, gql } from './_lib/apollo'

describe('Flow', () => {
  let postIdWithHiddenAtNull: string
  let postIdWithHiddenAtNotNull: string
  const userId1 = 'f38cc2ce-d100-459c-8859-4df77551b723'
  const userId2 = '81f9ea29-c8c4-4e13-b034-c66d991ad79b'

  beforeAll(async () => {
    // Insert post with hidden_at null
    const result1 = await apollo.mutate({
      mutation: gql`
        mutation InsertPostWithHiddenAtNull(
          $userId: uuid!
          $title: String!
          $content: String!
        ) {
          insert_post_one(
            object: {
              user_id: $userId
              title: $title
              content: $content
              hidden_at: null
            }
          ) {
            id
          }
        }
      `,
      variables: {
        userId: userId1,
        title: 'Visible Post',
        content: 'This post is visible.',
      },
      context: {
        headers: {
          'x-hasura-admin-secret': process.env.HASURA_GRAPHQL_ADMIN_SECRET,
        },
      },
    })
    postIdWithHiddenAtNull = result1.data.insert_post_one.id

    // Insert post with hidden_at not null
    const result2 = await apollo.mutate({
      mutation: gql`
        mutation InsertPostWithHiddenAtNotNull(
          $userId: uuid!
          $title: String!
          $content: String!
        ) {
          insert_post_one(
            object: {
              user_id: $userId
              title: $title
              content: $content
              hidden_at: "2024-08-16T00:00:00Z"
            }
          ) {
            id
          }
        }
      `,
      variables: {
        userId: userId1,
        title: 'Hidden Post',
        content: 'This post is hidden.',
      },
      context: {
        headers: {
          'x-hasura-admin-secret': process.env.HASURA_GRAPHQL_ADMIN_SECRET,
        },
      },
    })
    postIdWithHiddenAtNotNull = result2.data.insert_post_one.id
  })

  describe('Main test', () => {
    test('user can get post with hidden_at null', async () => {
      const result = await apollo.query({
        query: gql`
          query GetPostById($id: uuid!) {
            post_by_pk(id: $id) {
              id
              title
              content
              hidden_at
            }
          }
        `,
        variables: {
          id: postIdWithHiddenAtNull,
        },
        context: {
          headers: {
            'x-hasura-user-id': 'f38cc2ce-d100-459c-8859-4df77551b723',
            'x-hasura-role': 'user',
          },
        },
      })
      expect(result.data.post_by_pk).toBeDefined()
      expect(result.data.post_by_pk.hidden_at).toBeNull()
    })

    test('user cannot get post with hidden_at not null', async () => {
      const result = await apollo.query({
        query: gql`
          query GetPostById($id: uuid!) {
            post_by_pk(id: $id) {
              id
              title
              content
              hidden_at
            }
          }
        `,
        variables: {
          id: postIdWithHiddenAtNotNull,
        },
        context: {
          headers: {
            'x-hasura-user-id': 'f38cc2ce-d100-459c-8859-4df77551b723',
            'x-hasura-role': 'user',
          },
        },
      })
      expect(result.data.post_by_pk).toBeNull()
    })

    test('adds a post with the same user_id as X-HASURA-USER-ID', async () => {
      const userId = userId1
      const result = await apollo.mutate({
        mutation: gql`
          mutation AddPost($userId: uuid!, $title: String!, $content: String!) {
            insert_post_one(
              object: { user_id: $userId, title: $title, content: $content }
            ) {
              id
              title
              content
              user_id
            }
          }
        `,
        variables: {
          userId,
          title: 'Test Post',
          content: 'This is a test post.',
        },
        context: {
          headers: {
            'x-hasura-user-id': userId,
            'x-hasura-role': 'user',
          },
        },
      })
      expect(result.data.insert_post_one).toBeDefined()
      expect(result.data.insert_post_one.user_id).toBe(userId)
    })

    test('fails to add a post with different user_id and X-HASURA-USER-ID', async () => {
      const postUserId = userId1
      const hasuraUserId = userId2
      try {
        await apollo.mutate({
          mutation: gql`
            mutation AddPost(
              $userId: uuid!
              $title: String!
              $content: String!
            ) {
              insert_post_one(
                object: { user_id: $userId, title: $title, content: $content }
              ) {
                id
                title
                content
                user_id
              }
            }
          `,
          variables: {
            userId: postUserId,
            title: 'Test Post with Different User ID',
            content: 'This is a test post with a different user ID.',
          },
          context: {
            headers: {
              'x-hasura-user-id': hasuraUserId,
              'x-hasura-role': 'user',
            },
          },
        })
        // If the mutation does not throw an error, fail the test
        throw new Error('Expected mutation to fail, but it succeeded')
      } catch (error) {
        // Check if the error message matches the expected permission error
        expect(error.message).toContain(
          'check constraint of an insert/update permission has failed',
        )
      }
    })
  })
})
