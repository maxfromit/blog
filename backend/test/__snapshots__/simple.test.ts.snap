// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`First test > gets post 1`] = `
[
  {
    "__typename": "post",
    "id": "4c6ecaff-1908-4e82-a633-e27a2b18d594",
  },
  {
    "__typename": "post",
    "id": "51b5a629-a444-4835-816e-5c7feac099e0",
  },
  {
    "__typename": "post",
    "id": "ae348f93-300d-4878-a79b-34adf299eeba",
  },
  {
    "__typename": "post",
    "id": "ffa90e0f-7e39-4704-929f-1534b0907fc3",
  },
  {
    "__typename": "post",
    "id": "3b340a7b-9d0c-4598-9f52-fa36d5f2853b",
  },
]
`;
